src/base/static/components/input-form/index.js

```diff
@@ -0,0 +1,551 @@
+    );
+  }
+
+  componentWillReceiveProps(nextProps) {
+    if (nextProps.isContinuingFormSession) {
+      this.setActiveCategoryState(
+        this.getCategoryConfig(this.state.selectedCategory)
+      );
+
+      // TODO: I think this condition will no longer be necessary given reliable
+      // unmount behavior when the detail view renders.
+    } else if (nextProps.isLeavingForm) {
+      this.resetState();
+    }
+  }
+
+  componentWillUnmount() {
+    this.props.map.off("dragend", this.handleDragEnd);
+  }
+
+  getCategoryConfig(category) {
+    return this.props.placeConfig.place_detail.find(
+      config => config.category === category
+    );
+  }
+
+  onFieldChange(fieldName, fieldValue, isValid, message) {
+    const newState = this.state.fields
+      .get(fieldName)
+      .set(constants.FIELD_STATE_VALUE_KEY, fieldValue)
+      .set(constants.FIELD_STATE_VALIDITY_KEY, isValid)
+      .set(constants.FIELD_STATE_VALIDITY_MESSAGE_KET, message);
+    this.setState(({ fields, updatingField }) => ({
+      fields: fields.set(fieldName, newState),
+      updatingField: fieldName,
+    }));
+  }
+
+  onAdditionalData(action, payload) {
+    switch (action) {
+      case constants.ON_ADD_ATTACHMENT_ACTION:
+        this.attachments.push(payload);
+        break;
+      default:
+        console.error(
+          "Error: Unable to handle form field callback action:",
+          action
+        );
+        break;
+    }
+  }
+
+  setActiveCategoryState(categoryConfig) {
+    this.resetCategoryData();
+
+    let newFields = ImmutableMap();
+    categoryConfig.fields
+      .filter(field => {
+        if (field.type === constants.MAP_DRAWING_TOOLBAR_TYPENAME) {
+          this.customGeometryFieldname = field.name;
+        }
+
+        return (
+          this.resolveField(field).type !== constants.SUBMIT_FIELD_TYPENAME
+        );
+      })
+      .forEach(field => {
+        newFields = newFields.set(field.name, this.getInitialFieldState(field));
+        if (field.type === constants.RICH_TEXTAREA_FIELD_TYPENAME) {
+          this.richTextFields.push(field.name);
+        }
+      });
+
+    this.setState({
+      fields: newFields,
+      selectedCategory: categoryConfig.category,
+      isCategoryMenuCollapsed: true,
+      isFormSubmitting: false,
+      formValidationErrors: new Set(),
+      showValidityStatus: false,
+    });
+  }
+
+  getValidator(fieldConfig) {
+    if (fieldConfig.type === constants.SUBMIT_FIELD_TYPENAME) {
+      return {
+        validate: mayHaveAnyValue,
+        message: "",
+      };
+    } else if (fieldConfig.type === constants.MAP_DRAWING_TOOLBAR_TYPENAME) {
+      return {
+        validate: mustHaveSomeValue,
+        message: messages.missingGeometry,
+      };
+    } else {
+      return {
+        validate: fieldConfig.optional ? mayHaveAnyValue : mustHaveSomeValue,
+        message: messages.missingRequired,
+      };
+    }
+  }
+
+  getInitialFieldState(fieldConfig) {
```

Comment:
All of this method's logic seems like something that should be delegated to FormField. We already have a switch case there that initializes the fields. Can we put it there?

With the proper abstraction, I don't think our InputForm should have any references to specific form fields. Happy to discuss anytime.

Reply:
I'll take another pass at this-- I struggled with how best to set initial state in the top-level `input-form` component for all the various field types. If we delegate the above logic to fields, how can we set initial state in the form?

Reply:
Off the cuff, here's an idea: instead of holding a map in our `state.fields`, and having to support all of the logic around it in this component, `state.fields` could instead be an array of FormField models that handle that logic for us. The models can be an ES6 class, or perhaps the FormField itself. Then we delegate that logic to the FormField component as needed.

Happy to discuss further.

Reply:
I see. Or maybe the `field` component could just manually trigger an initial `onChange` event, which would set the state via the mechanism we already have in place.

Reply:
Sure, but using `onChange` for that purpose is bad practice. Perhaps you can add another prop like `onInitialize`.